language/libs/pointers.hpp

```diff
@@ -0,0 +1,378 @@
+ * when given a target pointer. I provide counted references to that slot
+ * that remain valid even after I have expired and cleared the stored
+ * target pointer.
+ */
+
+class SlotRef::Owner
+{
+public:
+    Owner() = default;
+    Owner(Owner const&) = delete;
+    Owner& operator =(Owner const&) = delete;
+
+    ~Owner() noexcept {
+        if (mRecord) {
+            mRecord->target = nullptr;
+            Registry::sRegistry.release(*mRecord);
+        }
+    }
+
+/**
+ * Occupy a registry slot with the given target pointer.
+ *
+ * @note This method must always be called with the same target pointer.
+ * @return A reference to the owned registry slot.
+ */
+
+    SlotRef get(void* const target)
+    {
+        if (!mRecord) {
+            mRecord = &Registry::sRegistry.take(target);
+            Registry::sRegistry.retain(*mRecord);
+        }
+        assert(target == mRecord->target);
+        return SlotRef{*mRecord};
+    }
+
+private:
+    Registry::Record* mRecord{nullptr};
+};
+
+
+/**
+ * I provide a typed pointer for a pointer stored in the registry.
+ *
+ * I become invalid when that entity expires.
+ */
+
+template<typename Value>
+class VolatilePtr
+{
+public:
+    constexpr VolatilePtr(SlotRef const& slotRef) noexcept
+    : mSlotRef{slotRef}
+    {}
+
+    constexpr Value* get() const noexcept
+    {
+        return static_cast<Value*>(mSlotRef->target);
+    }
+
+    constexpr Value& operator *() const noexcept
+    {
+        return *get();
+    }
+
+    constexpr Value* operator ->() const noexcept
+    {
+        return get();
+    }
+
+private:
+    SlotRef mSlotRef;
+};
+
+
+/**
+ * I hold an instance of Value.
+ *
+ * I can provide pointers which are invalidated when I have expired.
+ */
+
+template<typename Value>
+class Target
+{
+public:
+    struct forwardCtor {};
+
+/**
+ * Convenience function for invoking forwarding c'tor.
+ */
+
+    template<typename... Args>
+    static constexpr Target create(Args&&... args)
+    {
+        return Target(forwardCtor{}, std::forward<Args>(args)...);
+    }
+
+/**
+ * This constructor forwards its parameters to Value's c'tor, constructing
+ * the instance in-place.
+ */
+
+    template<typename... Args>
+    constexpr explicit Target(forwardCtor, Args&&... args)
+    : mValue{std::forward<Args>(args)...}
+    , mOwner{}
+    {}
+
+    Target(Target const& target) noexcept(std::is_nothrow_copy_constructible_v<Value>)
+    : mValue{target.mValue}
+    , mOwner{}
+    {}
+
+    Target(Target&& target) noexcept(std::is_nothrow_move_constructible_v<Value>)
+    : mValue{std::forward<decltype(target.mValue)>(target.mValue)}
+    , mOwner{}
+    {}
+
+    Target& operator =(Target const& target) noexcept(std::is_nothrow_copy_assignable_v<Value>)
+    {
+        mValue = target.mValue;
+        // Omit mOwner since it's a reference to ourselves.
+        return *this;
+    }
+
+    Target& operator =(Target&& target) noexcept(std::is_nothrow_move_assignable_v<Value>)
```

Comment:
That would make values wrapped inside `Target<>` behave like references:
```
auto a{Target<int>::create(0)};
auto b{Target<int>::create(1)};
b = a; // a and b are now 0
b = 1; // what is a now?
```
What you suggest would make `a == 1` at the end of this code. However, `Target<>` is intended to have value semantics.

Pointers provide the behaviour you describe:
```
auto a{Target<int>::create(0)};
auto b{Target<int>::create(1)};
auto aPtr{a.ptr()};
auto bPtr{b.ptr()};
bPtr = aPtr; // *aPtr and *bPtr both refer to a and are now 0
*bPtr = 1; // *aPtr and *bPtr both refer to a and are now 1
```

Reply:
I am suggesting it for move construction/assignment, not for copy construction/assignment...
Or would it be inconsistent and confusing then?

Reply:
```
auto a{Target<int>::create(0)};
auto b{Target<int>::create(1)};
b = a; // a and b are 0
*a = 1; // a is 1, b is 0
b = std::move(a); // a is indeterminate, b is 1
*a = 0; // a and b are 0
```
Yes, I think that's worse…

Reply:
But there is a rule that by no means is a programmer allowed to touch a variable after it has been moved (see here https://en.wikipedia.org/wiki/Touch-move_rule)...
And we can even enforce that by setting the a-pointing record to nullpointer when we move from it.

Reply:
Changing pointers in the registry changes all volatile pointers ever created for a target:
```
auto a{Target<int>::create(0)};
auto aPtr{a.ptr()};
auto b{Target<int>::create(1)};
b = a; // a and b are now 0, aPtr refers to a
*aPtr = 1; // a is 1, b is 0
b = std::move(a); // a is indeterminate, b is 1, aPtr refers to b (!)
*aPtr = 0; // a is indeterminate, b is 0 (!)
```
Furthermore, the important exception to that rule is that assigment is *always* allowed. If a target is only ever assigned to, then why waste effort changing what it refers to?

Reply:
Yeeeh yeeeh. I feel like I'm learning more in this discussion than you do, so who is the real winner in the end, hmm?
It still makes sense to me that if you move a into b for whatever reason, all pointers that point to a now point to b because b IS the new a.
Besides, check this out:
```cpp
std::vector<Target<Noisy>> noisies{};
noisies.emplace_back(Target<Noisy>::forwardCtor{}, "Vectorbert");

auto vecRef{noisies.front().ptr()};
std::cout << "vecRef=" << vecRef->name << '\n';

for (int i{0}; i != 4; ++i) {
	noisies.emplace_back(Target<Noisy>::forwardCtor{}, "goodbye");
}
std::cout << "vecRef=" << vecRef->name << '\n'; // BOOM! STILL VALID!
```
I argue that shooting yourself in the foot the way that is caused by moving a registry pointer along with its target hurts less than shooting into the std::vector moving something away from under my pointer foot. Although personal trauma may cause bias in my judgment here.

Reply:
Consider the definitions:
```
struct A {
};
struct B {
  B(B&&) {}
};
```
Given these definitions, `std::vector<A>` moves its elements during reallocation, while `std::vector<B>` copies. The reason is that since `B` is not `nothrow_move_constructible`, `std::vector<B>` must copy elements in order to satisfy the strong exception guarantee.

After successful reallocation, the old elements are released irrespective of whether they were copied or moved. This reliably clears any volatile pointers to those old elements, producing a reliable and non-surprising outcome independent of subtle properties of the element type.

Because these properties are so subtle, rather than dramatically changing behaviour based on copying v. moving, it's better to think of move semantics as an optimisation that is fickle and unreliable more often than not.

Reply:
Wellp, as the tribal wisdom of Dakota programmers teaches: When you discover you're riding a dead horse, the best strategy is to dismount.
I fought with all my prowess, but I must be judicious enough to recognize defeat. I shall cut my losses and lick my wounds. This round goes to you.

Reply:
The above suggestion won't work with the current registry design. In fact, `vecRef` on the last line will be **invalid** because it was cleared out when its target was destroyed.

Modifying pointers in the registry leads to the situation where several registry slots point to the same target. If each target only clears a single slot, pointers to the other slots are left dangling. To solve this, slots pointing to a target would need to be »chained«, leading to more complex code and significantly blowing up the registry.

The current design keeps it simple and efficient by enforcing that at most one slot points to each target.

Reply:
If you want to continue exploring the hypothetical scenario in which I didn't surrender yet:
We can swap the target pointers of the a-record and the b-record when someone writes
`a = move(b)` or `b = move(a)` :)

Reply:
Then all pointers to the destination will point to the source after the move, and vice versa. If you did *just* that, then the source clears the destination target's slot when it's destroyed, invalidating all pointers to the destination, and vice versa. Swapping the slot pointers in the associated targets is impossible with the current registry design because the slots refer to the value directly which doesn't need to be stored in a `Target` instance.

Reply:
> Then all pointers to the destination will point to the source after the move, and vice versa.

Yes, that's the point. If nobody touches the source again because you're not supposed to touch a moved-from thing then it's fine, no? I thought the whole point of move semantics is that the source is no longer used... But yes, the whole code would have to know when it was moved and if we require the programmer to program according to the rules then we don't need smart pointers.

> If you did just that, then the source clears the destination target's slot when it's destroyed, invalidating all pointers to the destination, and vice versa.

Hmm I see... Yes, we would also have to swap the owners.

> Swapping the slot pointers in the associated targets is impossible with the current registry design because the slots refer to the value directly which doesn't need to be stored in a Target instance.

I feel like here you anticipated "Hmm I see... Yes, we would also have to swap the owners." but I don't get it. I understand it like this:
A record points to a target, a target holds an owner of the slot/record, a volatilePtr holds a typed pointer to a record and nothing points to the value in the target directly.
I see that swapping is a bad idea (must have panicked) but couldn't we do the following:
`a = move(b)`
make the record that points to a point to b, make a the owner of b's slot/record, nullptrify the record that poins to old a (which invalidates all pointers that point to a but a has been overwritten and is gone and all these pointers have to accept the loss as if a went out of scope), make old a the owner of old b so the slot is cleaned up upon destruction.

Reply:
> Yes, that's the point. If nobody touches the source again because you're not supposed to touch a moved-from thing then it's fine, no? I thought the whole point of move semantics is that the source is no longer used...

That's fine for the source itself – but not for pointers to the source! As I said before, touching a moved-from value is *completely* ok. It is still required to be a valid object, it's just the value after being moved-from that is up to the implementation. Assigning to it, directly or through a pointer, is completely valid, and this would do something quite surprising if all pointers suddenly changed to the destination.

In general, changing a slot changes *all* volatile pointers behind their backs with no visible cue. Clearing the slot to `nullptr` is fine as that's the contract of a volatile pointer – making it point to something completely different is not part of that contract, and as it acts at a distance it's very confusing behaviour.

Reply:
> A record points to a target, a target holds an owner of the slot/record, a volatilePtr holds a typed pointer to a record and nothing points to the value in the target directly.

In the current implementation, records do point to the value directly: look at the pointer passed in `Target::ptr()`. Consequently, `VolatilePtr` doesn't need to know about `Target` but can just `static_cast` that pointer to its value type. Pointing to `Target` instead would bind these components more tightly without an obvious gain. On top of that, the registry ins't restricted to `Target` as its clients.

> make the record that points to a point to b, make a the owner of b's slot/record, nullptrify the record that poins to old a

There must be a misconception: if `a` is a `Target`, there is (at most) one record that points to and is owned by `a`. Changing it changes all pointers to `a`. There's no such thing as a »record that points to old a«.

> nullptrify the record that poins to old a (which invalidates all pointers that point to a but a has been overwritten and is gone and all these pointers have to accept the loss as if a went out of scope)

While it might be possible to evolve a new concept of »sticky pointers« from this, it's totally against how pointers work: pointers *refer* to variables (storage locations), and the fact that new values assigned to those locations can be observed through a pointer is precisely the point of it. So I'm not sure what you're getting at here. Changing the value of a variable is *not* the same as a variable going out of scope.

> make old a the owner of old b so the slot is cleaned up upon destruction.

Where does »old a« come from? There's only one storage for `a`.

Reply:
> As I said before, touching a moved-from value is completely ok.

Sorry, must have overread that in the heat of the moment.

> In the current implementation, records do point to the value directly

Oh right, the pointer is called target everywhere tho, it should be called value then, or targetvalue... Also, I confused the vloatilePtr -> record -> target chain with the record -> target -> value chain and I was thinking "no the volatilePtr doesn't point directly to the value, what's he writing about, what's the point of the whole construct then!?", my apologies. But that the record points to the value and not to the target does not cause a problem for what I wanted to do, does it? (I see all the other problems you pointed out, however.)

> There must be a misconception: if a is a Target, there is (at most) one record that points to and is owned by a. Changing it changes all pointers to a. There's no such thing as a »record that points to old a«.

Maybe it's better to close this discussion here and resolve any misunderstandings (also the ones I have with myself) on a hike.

> So I'm not sure what you're getting at here

Sticky pointers is a nice way to put it. I was thinking we can have a kind of bucket that can move through the memory and change the value in it but a pointer to it keeps pointing to it until it goes out of scope or is overwritten by another bucket. However, I was obviously not too aware of all the consequences.

Reply:
Screw the hike, here is what I meant:
![grafik](https://user-images.githubusercontent.com/12005608/61582991-64d66f00-ab32-11e9-9960-065e01158c79.png)
First we swap everything, then we nullpoint the record that points to old A (dashed line). But that was when I still thought it's fine when the B-record with the A-value is rubbish after this.

Reply:
But that leaves no way to access `A`, the destination of the assignment. And the effect of the assignment is only observable through A-Ptrs if you *don't* move the value to `A`.

While I don't think this can be made to work usefully for move-assignment, it is completely possible for move construction. In that case, there are no destination pointers or value to worry about, and the slot of the move-source can be taken over by the new target object.

Reply:
Aww! But that's already enough to make the vector thing work, which is everything I ever wanted! I'm glad we came to an agreement in the end.
By the way, at no point did I ever consider move assignment and move construction to be kind of the same thing, why would you even think that!

Reply:
It does work with `std::vector`, but is still sensitive to `std::is_nothrow_move_constructible` as described above, and can fail silently:
```auto const t0{Target<int>::create(123)};
auto const t1{std::move(t0)};```
performs copy-construction rather than move construction because `t0` is `const`.

Reply:
But should we still do it then? Although I like the feature, I don't know if the benefit in this one particular case is worth the fail-silently thing and also the inconsistency with move assignment...

Reply:
I think that decision is up to the language designer